File: hub/src/execute/effect_runner.py
import time
import copy
import platform

from light.effect.effect_priority_handler import EffectPriorityHandler
if platform.system() != 'Windows':
    from light.control.neopixel_driver import NeopixelDriver
from light.control.socketio_driver import SocketIoDriver

from light.effect.runtime_combiners.basic_combiner import BasicCombiner

class EffectRunner:
    def __init__(self, activator, sio):
        self.activator = activator
        self.priorityHandler = EffectPriorityHandler()
        self.sio = sio

    async def run(self):
        if platform.system() != 'Windows':
            driver = NeopixelDriver()
        else:
            driver = SocketIoDriver(self.sio)

        combiner = BasicCombiner()
        while (True):
            startTime = time.time()

            activeEffects = self.activator.getCurrentActiveEffects()

            priorityStartTime = time.time()
            self.priorityHandler.clear()
            for activeEffect in activeEffects:
                self.priorityHandler.addLedEffect(activeEffect)
            
            mappedEffectsToRun = self.priorityHandler.getLeds()
            #print("Priority: " + str(time.time() - priorityStartTime))

            combinerStart = time.time()
            combiner.combine(mappedEffectsToRun)
            #print("Combiner: " + str(time.time() - combinerStart))

            #print(mappedEffectsToRun.get(3))
            await driver.runEffects(mappedEffectsToRun)
            
            #print("Runner total: " + str(time.time() - startTime))
            #print("")

            await self.sio.sleep(0.01)

File: hub/drumlite-js/src/effect/EffectRunner.ts
import EffectPriorityHandler from "./EffectPriorityHandler";
import Util from "../util/Util";
import InstructionExecutor from "../driver/InstructionExecutor";
import IRemoteDriver from "../driver/IRemoteDriver";
import EffectActivator from "./EffectActivator";

export interface RunnerOptions {
    periodMillis?: number
}

export default class EffectRunner {
    activator: EffectActivator;
    priorityHandler: EffectPriorityHandler;
    ledDriver: IRemoteDriver;
    executor: InstructionExecutor;
    runLeds: boolean;
    options: RunnerOptions;

    constructor(activator: EffectActivator, driver: IRemoteDriver, options: RunnerOptions) {
        this.activator = activator;
        this.priorityHandler = new EffectPriorityHandler();
        this.ledDriver = driver;
        this.executor = new InstructionExecutor(this.ledDriver);
        this.runLeds = false;
        this.options = options;
    }

    public async run() {
        this.runLeds = true;
        while(this.runLeds) {
            const t = new Date().getTime();
            const activeEffects = this.activator.getCurrentActiveEffects();
            this.priorityHandler.clear();
            for (let activeEffect of activeEffects) {
                const instructions = activeEffect.getInstructions(t);
                instructions.map(instruction => this.priorityHandler.addLedEffect(instruction));
            }
            
            const mappedEffectsToRun = this.priorityHandler.getLeds();

            // TODO combiner or no combiner?
            this.executor.runEffects(mappedEffectsToRun)

            await Util.sleep(this.options.periodMillis || 10);
        }
    }

    public clear() {
        this.executor.clear();
    }

    public stop() {
        this.runLeds = false;
    }

    public setActivator(activator: EffectActivator) {
        this.activator = activator;
    }
}